# Evaluation criteria

The goal of this assignment is to get a view on your hands-on "data engineering" skills.  
At our company, our data scientists and engineers collaborate on projects.  
Your main focus will be creating performant & robust data flows.  
For a take-home-assignment, we cannot grant you access to our infrastructure.  
The assignement below measures your proficiency in general programming, data science & engineering tasks using python.  
Completion should not take more than half a day.

**We expect you to be proficient in:**
 * SQL queries (Sybase IQ system)
 * ETL flows (In collaboration with existing teams)
 * General python to glue it all together
 * Python data science ecosystem (Pandas + SKlearn)
 
**In this exercise we expect you to demonstrate your ability to / knowledge of:**
 * Building a data science runtime
 * PEP8 / Google python styleguide
 * Efficiently getting the job done
 * Choose meaningfull names for variables & functions
 * Writing maintainable code (yes, you might need to document some steps)
 * Help a data scientist present interactive results.
 * Offer predictions via REST api

# Setting-up a data science workspace

We allow you full freedom in setting up a data science runtime.  
The main objective is having a runtime where you can run this notebook and the code you will develop.  
You can choose for a local setup on your pc, or even a cloud setup if you're up for it.   

**In your environment, you will need things for:**
 * https request
 * python3 (not python2 !!)
 * (geo)pandas
 * interactive maps (e.g. folium, altair, ...)
 * REST apis
 
**Deliverables we expect**:
 * notebook with the completed assignment
 * list of packages for your runtime (e.g. yml or txt file)
 * evidence of a working API endpoint

# Importing packages

We would like you to put all your import statements here, together in 1 place.  
Before submitting, please make sure you remove any unused imports :-)  

In [15]:
## your imports go here.  You get pandas for free.
import pandas as pd
import requests
import joblib
import numpy as np
import unittest
import math


# Data ingestion exercises

## Getting store location data from an API

**Goal:** Obtain a pandas dataframe  
**Hint:** You will need to normalise/flatten the json, because it contains multiple levels  
**API call:** https://ecgplacesmw.colruytgroup.com/ecgplacesmw/v3/nl/places/filter/clp-places  

In [ ]:
## your code goes here

def get_clp_places(url):
    raise NotImplementedError("You need to develop this method!")

df_clp = get_clp_places("https://ecgplacesmw.colruytgroup.com/ecgplacesmw/v3/nl/places/filter/clp-places")
df_clp.head(10)

Solution Below:

In [16]:
url = "https://ecgplacesmw.colruytgroup.com/ecgplacesmw/v3/nl/places/filter/clp-places"
response  = requests.get(url)

if response.status_code == 200:
    data = response.json()
    df = pd.DataFrame(data)
    
df.head()

placeId                             ensign       commercialName branchId  \
0      902  {'id': 8, 'name': 'COLR_Colruyt'}      AALST (COLRUYT)     4156   
1      946  {'id': 8, 'name': 'COLR_Colruyt'}     AALTER (COLRUYT)     4218   
2      950  {'id': 8, 'name': 'COLR_Colruyt'}   AARSCHOT (COLRUYT)     4222   
3      886  {'id': 8, 'name': 'COLR_Colruyt'}  ALSEMBERG (COLRUYT)     4138   
4      783  {'id': 8, 'name': 'COLR_Colruyt'}       AMAY (COLRUYT)     3853   

  sourceStatus                                          placeType  \
0           AC  {'id': 1, 'longName': 'Winkel', 'placeTypeDesc...   
1           AC  {'id': 1, 'longName': 'Winkel', 'placeTypeDesc...   
2           AC  {'id': 1, 'longName': 'Winkel', 'placeTypeDesc...   
3           AC  {'id': 1, 'longName': 'Winkel', 'placeTypeDesc...   
4           AC  {'id': 1, 'longName': 'Winkel', 'placeTypeDesc...   

       sellingPartners                  handoverServices  \
0  [QUALITY, 3RDPARTY]  [CSOP_ORDERABLE, PREPAID_PARCEL]   
1  [QUALITY, 3RDPARTY]  [CSOP_ORDERABLE, PREPAID_PARCEL]   
2  [QUALITY, 3RDPARTY]  [CSOP_ORDERABLE, PREPAID_PARCEL]   
3  [QUALITY, 3RDPARTY]  [CSOP_ORDERABLE, PREPAID_PARCEL]   
4  [QUALITY, 3RDPARTY]  [CSOP_ORDERABLE, PREPAID_PARCEL]   

                                     geoCoordinates  \
0   {'latitude': 50.933074, 'longitude': 4.0538972}   
1  {'latitude': 51.0784761, 'longitude': 3.4500133}   
2  {'latitude': 50.9760369, 'longitude': 4.8110969}   
3   {'latitude': 50.7415212, 'longitude': 4.336719}   
4  {'latitude': 50.5599284, 'longitude': 5.3061951}   

                                             address  \
0  {'streetName': 'BRUSSELSE STEENWEG', 'houseNum...   
1  {'streetName': 'LOSTRAAT', 'houseNumber': '66'...   
2  {'streetName': 'LEUVENSESTEENWEG', 'houseNumbe...   
3  {'streetName': 'BRUSSELSESTEENWEG', 'houseNumb...   
4  {'streetName': 'CHAUSSEE DE TONGRES', 'houseNu...   

                                         moreInfoUrl  \
0  https://www.colruyt.be/nl/colruyt-openingsuren...   
1  https://www.colruyt.be/nl/colruyt-openingsuren...   
2  https://www.colruyt.be/nl/colruyt-openingsuren...   
3  https://www.colruyt.be/nl/colruyt-openingsuren...   
4  https://www.colruyt.be/nl/colruyt-openingsuren...   

                                            routeUrl  isActive  \
0  https://maps.apple.com/?daddr=50.933074,4.0538972      True   
1  https://maps.apple.com/?daddr=51.0784761,3.450...      True   
2  https://maps.apple.com/?daddr=50.9760369,4.811...      True   
3  https://maps.apple.com/?daddr=50.7415212,4.336719      True   
4  https://maps.apple.com/?daddr=50.5599284,5.306...      True   

                             placeSearchOpeningHours  
0  [{'date': '12-08-2023', 'opens': 830, 'closes'...  
1  [{'date': '12-08-2023', 'opens': 830, 'closes'...  
2  [{'date': '12-08-2023', 'opens': 830, 'closes'...  
3  [{'date': '12-08-2023', 'opens': 830, 'closes'...  
4  [{'date': '12-08-2023', 'opens': 830, 'closes'...

In [17]:
df_v2= df.copy()

In [18]:
df_v2= df.copy()
cols_to_flatten = ['ensign','placeType','geoCoordinates','address','placeSearchOpeningHours']
flat_df = []

for i in cols_to_flatten:
    
    # Apply the flatten_dict function to each row in the DataFrame
    temp_df = pd.json_normalize(df[i])

    # Create a new DataFrame with the flattened data
    flat_df = pd.DataFrame(temp_df)
    
    df_v2 = pd.concat([df_v2,flat_df],axis=1)
    df_v2.drop([i],inplace=True, axis=1)

### Quality checks

We would like you to add several checks on this data based on these constraints:  
 * records > 200
 * latitude between 49 and 52
 * longitude between 2 and 7
 
We dont want you to create a full blown test suite here, we're just gonna use 'asserts' from unittest

Solution Below

In [19]:
##Checks on data based on above three constraints
tc = unittest.TestCase('__init__')

#tc.assertEquals(1, 2, "Replace this assert with meaningfull checks")

tc.assertEquals(len(df_v2)> 200, True, "Records > 200")

tc.assertEquals(all(49 <= lat <= 52 for lat in df_v2['latitude']), True, "latitude between 49 and 52")

tc.assertEquals(all(2 <= lon <= 7 for lon in df_v2['longitude']), True, "longitude between 2 and 7")

C:\Users\N-Workstation\AppData\Local\Temp\ipykernel_11564\1832540710.py:6: DeprecationWarning: Please use assertEqual instead.
  tc.assertEquals(len(df_v2)> 200, True, "Records > 200")
C:\Users\N-Workstation\AppData\Local\Temp\ipykernel_11564\1832540710.py:8: DeprecationWarning: Please use assertEqual instead.
  tc.assertEquals(all(49 <= lat <= 52 for lat in df_v2['latitude']), True, "latitude between 49 and 52")
C:\Users\N-Workstation\AppData\Local\Temp\ipykernel_11564\1832540710.py:10: DeprecationWarning: Please use assertEqual instead.
  tc.assertEquals(all(2 <= lon <= 7 for lon in df_v2['longitude']), True, "longitude between 2 and 7")


### Feature creation

Create a new column "antwerpen" which is 1 for all stores in Antwerpen (province) and 0 for all others 

In [ ]:
## your code goes here
raise NotImplementedError("You need to develop this feature!")

df_clp["antwerpen"].value_counts()

Solution:

In [20]:
df_v2['antwerpen'] = df_v2['cityName'].str.contains('Antwerpen', case=False).astype(int)

In [21]:
df_v2["antwerpen"].value_counts()

0    252
1      7
Name: antwerpen, dtype: int64

## Predict used car value

A datascientist in our team made a basic model to predict car prices.  
The model was saved to disk ('lgbr_cars.model') using joblib's dump fuctionality.  
Documentation states the model is a LightGBM Regressor, trained using the sk-learn api.  

**As engineer, your task it to expose this model as REST-api.** 

First, retrieve the model via the function below.  
Change the path according to your setup.  

In [ ]:
## your code goes here

def retrieve_model(path):
    raise NotImplementedError("You need to retrieve the trained model!")
    return trained_model

lgbr_cars = retrieve_model("../../models/lgbr_cars.model")

tc.assertEqual(str(type(lgbr_cars)),"<class 'lightgbm.sklearn.LGBMRegressor'>", type(lgbr_cars))

Solution:

In [22]:

tc = unittest.TestCase('__init__')

# Load the saved model from the file
lgbr_cars = joblib.load('lgbr_cars.model')

tc.assertEqual(str(type(lgbr_cars)),"<class 'lightgbm.sklearn.LGBMRegressor'>", type(lgbr_cars))

Now you have your trained model, lets do a functional test based on the parameters below.  
You have to present the parameters in this order.  

* vehicleType: coupe
* gearbox: manuell
* powerPS: 190
* model: NaN
* kilometer: 125000
* monthOfRegistration: 5 
* fuelType: diesel
* brand: audi

Based on these parameters, you should get a predicted value of 14026.35068804
However, the model doesnt accept string inputs, see the integer encoding below:

In [23]:
model_test_input = [[3,1,190,-1,125000,5,3,1]]

In [24]:
## your code goes here

def make_prediction(trained_model, single_input):
    predicted_value =  trained_model.predict(single_input)
    #raise NotImplementedError("You need to predict the value!")
    return predicted_value

predicted_value = make_prediction(lgbr_cars, model_test_input)
#print(np.round(predicted_value,2))

tc.assertAlmostEqual(np.round(predicted_value,2), 14026.35, places=2)

In [25]:
model_test_input_RestApi = [[-1,1,0,118,150000,0,1,38]]

predicted_value = make_prediction(lgbr_cars, model_test_input_RestApi)

print(predicted_value)

[13920.70463564]


Now you got this model up and running, we want you to **expose it as a rest api.**  
We don't expect you to set up any authentication.  
We're not looking for beautiful inputs, just make it work.  
**Building this endpoint should NOT be done in a notebook, but in proper .py file(s)**

Once its up and running, use it to predict the following input:
* [-1,1,0,118,150000,0,1,38] ==> prediction should be 13920.70

***Solution:***
    
Please refer attached app.py file & please execute python app.py in your terminal to start flask app & you can use curl (tested on gitbash) to test the endpoint: curl -X POST -H "Content-Type: application/json" -d '{"data": [-1,1,0,118,150000,0,1,38]}' http://127.0.0.1:5000/predict


## Geospatial data exercise
The goal of this exercise is to read in some data from a shape file and visualize it on a map
- The map should be dynamic. I want to zoom in and out to see more interesting aspects of the map
- We want you to visualize the statistical sectors within a distance of 2KM of your home location.

Specific steps to take:
- Read in the shape file
- Transform to WGS coordinates
- Create a distance function (Haversine)
- Create variables for home_lat, home_lon and perimeter_distance
- Calculate centroid for each nis district
- Calculate the distance to home for each nis district centroid 
- Figure out which nis districts are near your home
- Create dynamic zoomable map
- Visualize the nis districts near you (centroid <2km away), on the map


In [1]:
# Some imports to help you along the way
import geopandas as gpd
import folium # you can use any viz library you prefer

In [ ]:
# part 1: Reading in the data
# get this file from https://statbel.fgov.be/sites/default/files/files/opendata/Statistische%20sectoren/sh_statbel_statistical_sectors_20200101.shp.zip 
df = gpd.read_file('AD_0_StatisticSector.shp')
df = df.to_crs({'init': 'epsg:4326'}) # change projection to wgs84 

# One of the data scientists discovered stackoverflow ;-) and copypasted something from https://gis.stackexchange.com/questions/166820/geopandas-return-lat-and-long-of-a-centroid-point
# A data science engineer should be able to speed this next code up

for i in range(0, len(df)):
    df.loc[i,'centroid_lon'] = df.geometry.centroid.x.iloc[i]
    df.loc[i,'centroid_lat'] = df.geometry.centroid.y.iloc[i]
    
    


Solution Starts below:

In [2]:
gdf = gpd.read_file('sh_statbel_statistical_sectors_20200101.shp')
gdf = gdf.to_crs({'init': 'epsg:4326'}) # change projection to wgs84 

# Calculate centroids in the projected CRS
gdf['centroid_lon'] = gdf.geometry.centroid.x
gdf['centroid_lat'] = gdf.geometry.centroid.y

c:\ProgramData\anaconda3\envs\dse_candidate\lib\site-packages\pyproj\crs\crs.py:141: FutureWarning: '+init=<authority>:<code>' syntax is deprecated. '<authority>:<code>' is the preferred initialization method. When making the change, be mindful of axis order changes: https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6
  in_crs_string = _prepare_from_proj_string(in_crs_string)


In [5]:
gdf.head()

CS01012020            T_SEC_NL            T_SEC_FR            T_SEC_DE  \
0  11001A00-  AARTSELAAR-CENTRUM  AARTSELAAR-CENTRUM  AARTSELAAR-CENTRUM   
1  11001A01-        DE LEEUWERIK        DE LEEUWERIK        DE LEEUWERIK   
2  11001A020           BUERSTEDE           BUERSTEDE           BUERSTEDE   
3  11001A030            YSSELAAR            YSSELAAR            YSSELAAR   
4  11001A042       KLEINE GRIPPE       KLEINE GRIPPE       KLEINE GRIPPE   

   C_NIS6   T_NIS6_NL   T_NIS6_FR CNIS5_2020    T_MUN_NL    T_MUN_FR  ...  \
0  11001A  AARTSELAAR  AARTSELAAR      11001  Aartselaar  Aartselaar  ...   
1  11001A  AARTSELAAR  AARTSELAAR      11001  Aartselaar  Aartselaar  ...   
2  11001A  AARTSELAAR  AARTSELAAR      11001  Aartselaar  Aartselaar  ...   
3  11001A  AARTSELAAR  AARTSELAAR      11001  Aartselaar  Aartselaar  ...   
4  11001A  AARTSELAAR  AARTSELAAR      11001  Aartselaar  Aartselaar  ...   

         T_REGIO_DE C_COUNTRY NUTS1 NUTS2  NUTS3  M_AREA_HA M_PERI_M  \
0  Flämische Region        BE   BE2  BE21  BE211  53.158103   5248.0   
1  Flämische Region        BE   BE2  BE21  BE211  67.114679   3935.0   
2  Flämische Region        BE   BE2  BE21  BE211  28.084365   2202.0   
3  Flämische Region        BE   BE2  BE21  BE211  42.811849   3403.0   
4  Flämische Region        BE   BE2  BE21  BE211  25.561955   2166.0   

                                            geometry centroid_lon centroid_lat  
0  POLYGON Z ((4.39247 51.13717 0.00000, 4.39519 ...     4.388756    51.132421  
1  POLYGON Z ((4.39374 51.13858 0.00000, 4.39356 ...     4.384630    51.138929  
2  POLYGON Z ((4.37785 51.13689 0.00000, 4.37634 ...     4.374922    51.139997  
3  POLYGON Z ((4.39140 51.14451 0.00000, 4.39057 ...     4.382041    51.144913  
4  POLYGON Z ((4.39374 51.13858 0.00000, 4.39350 ...     4.391940    51.142112  

[5 rows x 32 columns]

In [18]:
# Let's create some variables to indicate the location of your interest 
home_lat = 51.132421
home_lon = 4.288756
perimeter_distance = 2 # km

In [ ]:
# At some point we will need a distance function (google the Haversine formula, and implement it)

def haversine(lat1, lon1, lat2, lon2):
    raise NotImplementedError('Implement haversine')


In [5]:
import math
def haversine_distance(lat1, lon1, lat2, lon2):
    # Radius of the Earth in kilometers
    R = 6371.0

    # Convert latitude and longitude from degrees to radians
    lat1_rad = math.radians(lat1)
    lon1_rad = math.radians(lon1)
    lat2_rad = math.radians(lat2)
    lon2_rad = math.radians(lon2)

    # Differences between the latitudes and longitudes
    dlat = lat2_rad - lat1_rad
    dlon = lon2_rad - lon1_rad

    # Haversine formula
    a = math.sin(dlat / 2)**2 + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon / 2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = R * c

    return distance

Next, implement some sanity checks for your distance function 

In [7]:

# implement sanity checks here
import unittest
tc = unittest.TestCase('__init__')

#Checking for distance is zero
tc.assertAlmostEqual(haversine_distance(52.5200, 13.4050, 52.5200, 13.4050), 0)

Now, create a dynamical map 

In [ ]:
# implementation of the map goes here

Solution:

In [11]:
gdf.head(2)

CS01012020            T_SEC_NL            T_SEC_FR            T_SEC_DE  \
0  11001A00-  AARTSELAAR-CENTRUM  AARTSELAAR-CENTRUM  AARTSELAAR-CENTRUM   
1  11001A01-        DE LEEUWERIK        DE LEEUWERIK        DE LEEUWERIK   

   C_NIS6   T_NIS6_NL   T_NIS6_FR CNIS5_2020    T_MUN_NL    T_MUN_FR  ...  \
0  11001A  AARTSELAAR  AARTSELAAR      11001  Aartselaar  Aartselaar  ...   
1  11001A  AARTSELAAR  AARTSELAAR      11001  Aartselaar  Aartselaar  ...   

         T_REGIO_DE C_COUNTRY NUTS1 NUTS2  NUTS3  M_AREA_HA M_PERI_M  \
0  Flämische Region        BE   BE2  BE21  BE211  53.158103   5248.0   
1  Flämische Region        BE   BE2  BE21  BE211  67.114679   3935.0   

                                            geometry centroid_lon centroid_lat  
0  POLYGON Z ((4.39247 51.13717 0.00000, 4.39519 ...     4.388756    51.132421  
1  POLYGON Z ((4.39374 51.13858 0.00000, 4.39356 ...     4.384630    51.138929  

[2 rows x 32 columns]

In [21]:
center_lat, center_lon = gdf.geometry.centroid.y.mean(), gdf.geometry.centroid.x.mean()

# Create the base map centered around your location
m = folium.Map(location=[center_lat, center_lon], zoom_start=10)

# Add your home marker
folium.Marker([home_lat, home_lon], popup="Home", icon=folium.Icon(color="red")).add_to(m)

# Add markers for nearby districts
for index, row in gdf.iterrows():
    district_lat = row["centroid_lat"]
    district_lon = row["centroid_lon"]
    distance = haversine_distance(home_lat, home_lon,district_lat, district_lon)
    if distance <= perimeter_distance:
        folium.Marker([district_lat, district_lon], popup=row["T_SEC_NL"]).add_to(m)

# Display the map
m.save("nic_districts_map.html")


Once the above map gets created.. you can open the html file & looks at the places under 2 KM from HOME location coorinates